pep-0634.rst

```diff
@@ -0,0 +1,696 @@
+
+
+Syntax and Semantics
+====================
+
+TODO: Should we show the lookaheads in the grammar?  They are a bit
+ugly.  Maybe as a compromise show them in the appendix but not in the
+main text?
+
+See `Appendix A`_ for the complete grammar.
+
+
+The ``match`` statement
+-----------------------
+
+A ``match`` statement has the following top-level grammar::
+
+  match_stmt: "match" match_expr ':' NEWLINE INDENT case_block+ DEDENT
+  match_expr:
+      | star_named_expression ',' star_named_expressions?
+      | named_expression
+  case_block: "case" patterns [guard] ':' block
+  guard: 'if' named_expression
+
+The rules ``star_named_expression``, ``star_named_expressions``,
+``named_expression`` and ``block`` are part of the `standard Python
+grammar <https://docs.python.org/3.10/reference/grammar.html>`_.
+
+The rule ``patterns`` is specified below.
+
+For context, ``match_stmt`` is a new alternative for
+``compound_statement``::
+
+  compound_statement:
+      | if_stmt
+      ...
+      | match_stmt
+
+
+The ``match`` and ``case`` keywords are soft keywords, i.e. they are
+not reserved words in other grammatical contexts (including at the
+start of a line if there is no colon where expected).  This implies
+that they are recognized as keywords when part of a ``match``
+statement or ``case`` block only, and are allowed to be used in all
+other context as variable or argument names.
+
+
+Match semantics
+~~~~~~~~~~~~~~~
+
+TODO: Make the language about choosing a block more precise.
+
+The overall semantics for choosing the match is to choose the first
+matching pattern (including guard) and execute the corresponding
+block.  The remaining patterns are not tried.  If there are no
+matching patterns, execution continues at the following statement.
+
+Name bindings made during a successful pattern match outlive the
+executed block and can be used after the match statement.
+
+During failed pattern matches, some sub-patterns may succeed. For
+example, while matching the value ``[0, 1, 2]`` with the pattern ``(0,
+x, 1)``, the sub-pattern ``x`` may succeed if the list elements are
+matched from left to right.  The implementation may choose to either
+make persistent bindings for those partial matches or not. User code
+including a ``match`` statement should not rely on the bindings being
+made for a failed match, but also shouldn't assume that variables are
+unchanged by a failed match.  This part of the behavior is left
+intentionally unspecified so different implementations can add
+optimizations, and to prevent introducing semantic restrictions that
+could limit the extensibility of this feature.
+
+The precise pattern binding rules vary per pattern type and are
+specified below.
+
+
+.. _patterns:
+
+Patterns
+--------
+
+TODO: rewrite the ``patterns`` rule to be easier to follow -- why
+isn't it just ``','.value_pattern+ [',']``?  Also, ``value_pattern``
+is a confusing name, since it is unrelated to "constant value
```

Comment:
It's this way because you can't have a starred pattern without a comma.

Reply:
Hm, then I still propose to give it a better name.

Reply:
Hm. `elts_pattern` and `elt_pattern` instead of `values_pattern` and `value_pattern`? `items_pattern` is already taken by mappings.

Reply:
Let me ponder that, I'm not crazy about the abbreviated name. Maybe `multi_pattern`? Also, could we define the name `open_pattern` to mean `multi_pattern | pattern`? It somewhat nicely contrasts with `closed_pattern`. (Although then apparently walrus and OR patterns are neither closed nor open. :-)